code/modules/antagonists/changeling/changeling.dm

```diff
@@ -74,324 +102,442 @@
+	if(!living_source.mind)
 		return
 
-	INVOKE_ASYNC(chosen_sting, /datum/action/changeling/sting.proc/try_to_sting, ling, A)
+	regain_powers()
 
-	return COMSIG_MOB_CANCEL_CLICKON
+/*
+ * Signal proc for [COMSIG_LIVING_LIFE].
+ * Handles regenerating chemicals on life ticks.
+ */
+/datum/antagonist/changeling/proc/on_life(datum/source, delta_time, times_fired)
+	SIGNAL_HANDLER
 
-/datum/antagonist/changeling/proc/has_sting(datum/action/changeling/power)
-	for(var/P in purchasedpowers)
-		var/datum/action/changeling/otherpower = P
-		if(initial(power.name) == otherpower.name)
-			return TRUE
-	return FALSE
+	if(!iscarbon(owner.current))
+		return
 
+	// If dead, we only regenerate up to half chem storage.
+	if(owner.current.stat == DEAD)
+		adjust_chemicals((chem_recharge_rate - chem_recharge_slowdown) * delta_time, total_chem_storage * 0.5)
 
-/datum/antagonist/changeling/proc/purchase_power(sting_name)
-	var/datum/action/changeling/thepower
+	// If we're not dead - we go up to the full chem cap.
+	else
+		adjust_chemicals((chem_recharge_rate - chem_recharge_slowdown) * delta_time)
 
-	for(var/path in all_powers)
-		var/datum/action/changeling/S = path
-		if(initial(S.name) == sting_name)
-			thepower = new path
-			break
+/*
+ * Signal proc for [COMSIG_MOB_MIDDLECLICKON] and [COMSIG_MOB_ALTCLICKON].
+ * Allows the changeling to sting people with a click.
+ */
+/datum/antagonist/changeling/proc/on_click_sting(mob/living/carbon/ling, atom/clicked)
+	SIGNAL_HANDLER
 
-	if(!thepower)
-		to_chat(owner.current, "This is awkward. Changeling power purchase failed, please report this bug to a coder!")
+	if(!chosen_sting || clicked == ling || !istype(ling) || ling.stat != CONSCIOUS)
 		return
 
-	if(absorbedcount < thepower.req_dna)
-		to_chat(owner.current, span_warning("We lack the energy to evolve this ability!"))
+	INVOKE_ASYNC(chosen_sting, /datum/action/changeling/sting.proc/try_to_sting, ling, clicked)
+
+	return COMSIG_MOB_CANCEL_CLICKON
+
+/*
+ * Adjust the chem charges of the ling by [amount]
+ * and clamp it between 0 and override_cap (if supplied) or total_chem_storage (if no override supplied)
+ */
+/datum/antagonist/changeling/proc/adjust_chemicals(amount, override_cap)
+	if(!isnum(amount))
 		return
+	var/cap_to = isnum(override_cap) ? override_cap : total_chem_storage
+	chem_charges = clamp(chem_charges + amount, 0, cap_to)
+
+	owner.current.hud_used?.lingchemdisplay.maptext = FORMAT_CHEM_CHARGES_TEXT(chem_charges)
+
+/*
+ * Remove changeling powers from the current Changeling's purchased_powers list.
+ *
+ * if [include_innate] = TRUE, will also remove all powers from the Changeling's innate_powers list.
+ */
+/datum/antagonist/changeling/proc/remove_changeling_powers(include_innate = FALSE)
+	if(!isliving(owner.current))
+		return
+
+	if(chosen_sting)
+		chosen_sting.unset_sting(owner.current)
+
+	for(var/datum/action/changeling/power as anything in purchased_powers)
+		purchased_powers -= power
+		power.Remove(owner.current)
```

Comment:
This I think should just happen in action qdel. I think you could replace this whole forloop with just a QDEL_LIST macro